ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1881,18 +2057,318 @@ void shutdown() throws InterruptedException {
+            maxReplicaCountPerRequest > 0 ? Utils.partitionList(remoteReplicasPerNode, maxReplicaCountPerRequest)
+                : Collections.singletonList(remoteReplicasPerNode);
+
+        // for each of smaller array of remote replicas update with current group id and add to groupId map
+        for (List<RemoteReplicaInfo> remoteReplicaInfos : remoteReplicaSegregatedList) {
+          groupIdToRemoteReplicaMap.put(groupId, remoteReplicaInfos);
+          logger.trace("Thread name: {} Assigned groupId {} to replicas {}", threadName, groupId, remoteReplicaInfos);
+          groupId++;
+        }
+
+        // now, for this data node add one stand by group id for standby remote replicas
+        standByGroupIds.put(remoteHost, groupId);
+        logger.trace("Thread name: {} Assigned groupId for standBy groups {} to remoteHost {}", threadName, groupId,
+            remoteHost);
+        groupId++;
+      }
+    }
+
+    /**
+     * Removes finished groups from inflight remote replica groups  {@link #inflightGroups}
+     * drops cached keys from storeKeyConverter.
+     * For finished groups adds time until which replicas are throttled in to {@link #replicaToThrottledTill}
+     */
+    private void processFinishedGroups() {
+      Iterator<Map.Entry<Integer, RemoteReplicaGroup>> inflightRemoteReplicaGroupIterator =
+          inflightGroups.entrySet().iterator();
+
+      while (inflightRemoteReplicaGroupIterator.hasNext()) {
+        Map.Entry<Integer, RemoteReplicaGroup> entry = inflightRemoteReplicaGroupIterator.next();
+        RemoteReplicaGroup remoteReplicaGroup = entry.getValue();
+
+        if (remoteReplicaGroup.isDone()) {
+          long currentTime = time.milliseconds();
+          long throttledTill = currentTime + threadThrottleDurationMs;
+          remoteReplicaGroup.remoteReplicaInfos.forEach(remoteReplicaInfo -> {
+            replicaToThrottledTill.put(remoteReplicaInfo, throttledTill);
+            logger.trace("Thread name: {} Remote Replica {} is throttled till {} in replica", threadName,
+                remoteReplicaInfo, throttledTill);
+          });
+          storeKeyConverter.tryDropCache(remoteReplicaGroup.getStoreKeysConversionLog());
+          inflightRemoteReplicaGroupIterator.remove();
+          logger.trace("Thread name: {} Group Id {} removed from inflight groups", threadName, entry.getKey());
+        }
+      }
+
+      logger.trace("Thread name: {} Finished processing finished groups", threadName);
+    }
+
+    /**
+     * checks if any signal to terminate cycle is received,
+     * checks whether any group has reached max iteration count
+     * @return {@link boolean} returns true if new groups can be created, false otherwise
+     */
+    private boolean shouldCreateNewGroups() {
+      // check if thread is in shutdown or cycle termination is received or max iterations reached
+      if (!running || shouldTerminateCurrentCycle) {
```

Comment:
Let's add more comments here, to list out the conditions when we don't want to create new groups.
1. the process is shutting down, so no need to create new group even if we have more iteration for them to run.
2. shouldTerminateCurrentCycle is true, which means there are replicas added to this thread or removed from this thread, don't create new group and just let existing groups finish so we can later create new groups with the new set of replicas in next cycle.
3. any group reaches max iteration, which means existing groups should just finish its current work and stop, thus no new groups should be created.

Reply:
Added comments